mlflow/server/js/src/experiment-tracking/components/ParallelCoordinatesPlotPanel.js

```diff
@@ -25,36 +25,126 @@ export class ParallelCoordinatesPlotPanel extends React.Component {
     sharedParamKeys: PropTypes.arrayOf(String).isRequired,
     // An array of metric keys shared by all runs
     sharedMetricKeys: PropTypes.arrayOf(String).isRequired,
+    // An array of different parameter keys across runs
+    diffParamKeys: PropTypes.arrayOf(String).isRequired,
   };
 
-  state = {
-    // Default to select all parameters
-    selectedParamKeys: this.props.sharedParamKeys,
-    // Default to select the first metric key.
-    // Note that there will be no color scaling if no metric is selected.
-    selectedMetricKeys: this.props.sharedMetricKeys.slice(0, 1),
+  constructor(props) {
+    super(props);
+    const { sharedParamKeys, sharedMetricKeys } = props;
+
+    this.state = {
+      // Default to select all parameters
+      selectedParamKeys: sharedParamKeys,
+      // Default to select the first metric key.
+      // Note that there will be no color scaling if no metric is selected.
+      selectedMetricKeys: sharedMetricKeys.slice(0, 1),
+      selectAll: this.getSelectAllState(sharedParamKeys),
+      selectDiff: this.getSelectDiffState(sharedParamKeys),
+    };
+  }
+
+  getSelectAllState = (selectedParamKeys) => {
+    const { allParamKeys } = this.props;
+
+    const allParamsSelected = _.isEqual(selectedParamKeys, allParamKeys);
+    return {
+      indeterminate: !allParamsSelected && selectedParamKeys.length > 0,
+      checked: allParamsSelected,
+    };
+  };
+
+  getSelectDiffState = (selectedParamKeys) => {
+    const { diffParamKeys } = this.props;
+
+    const allDiffParamsSelected = _.isEqual(selectedParamKeys, diffParamKeys);
+    const diffParamsSelected = _.intersection(selectedParamKeys, diffParamKeys).length > 0;
+    return {
+      indeterminate: !allDiffParamsSelected && diffParamsSelected,
+      checked: allDiffParamsSelected,
+      disabled: diffParamKeys.length === 0,
+    };
   };
 
   handleParamsSelectChange = (paramValues) => {
-    this.setState({ selectedParamKeys: paramValues });
+    this.setState({
+      selectedParamKeys: paramValues,
+      selectAll: this.getSelectAllState(paramValues),
+      selectDiff: this.getSelectDiffState(paramValues),
+    });
   };
 
   handleMetricsSelectChange = (metricValues) => {
     this.setState({ selectedMetricKeys: metricValues });
   };
 
+  handleSelectAllChange = () => {
```

Comment:
I expected this method to just say

if-checked, set selectedParamKeys=[] otherwise selectedParamKeys= allParamKeys.

I think there should be some other method which recomputes the selectAll/selectDiff derived values. For example, what if `this.selectAllState` was just a method that read `selectedParamKeys` from the current state?

Reply:
That makes sense. I'll fix this.